subsys/bluetooth/mesh/lightness_srv.c

```diff
@@ -0,0 +1,805 @@
+		      struct bt_mesh_msg_ctx *ctx,
+		      const struct bt_mesh_onoff_set *onoff_set,
+		      struct bt_mesh_onoff_status *rsp)
+{
+	struct bt_mesh_lightness_srv *srv = CONTAINER_OF(
+		onoff_srv, struct bt_mesh_lightness_srv, ponoff.onoff);
+	struct bt_mesh_lightness_set set = {
+		.transition = onoff_set->transition,
+	};
+	struct bt_mesh_lightness_status status;
+
+	if (onoff_set->on_off) {
+		set.lvl = (srv->default_light ? srv->default_light : srv->last);
+	} else {
+		set.lvl = 0;
+	}
+
+	change_lvl(srv, ctx, &set, &status);
+
+	if (rsp) {
+		rsp->present_on_off = (status.current > 0);
+		rsp->remaining_time = status.remaining_time;
+		rsp->target_on_off = (status.target > 0);
+	}
+}
+
+static void onoff_get(struct bt_mesh_onoff_srv *onoff_srv,
+		      struct bt_mesh_msg_ctx *ctx,
+		      struct bt_mesh_onoff_status *rsp)
+{
+	struct bt_mesh_lightness_srv *srv = CONTAINER_OF(
+		onoff_srv, struct bt_mesh_lightness_srv, ponoff.onoff);
+	struct bt_mesh_lightness_status status = { 0 };
+
+	srv->handlers->light_get(srv, ctx, &status);
+
+	rsp->present_on_off = (status.current > 0);
+	rsp->remaining_time = status.remaining_time;
+	rsp->target_on_off = (status.target > 0);
+}
+
+const struct bt_mesh_onoff_srv_handlers
+	_bt_mesh_lightness_srv_onoff_handlers = {
+		.set = onoff_set,
+		.get = onoff_get,
+	};
+
+static void bt_mesh_lightness_srv_reset(struct bt_mesh_model *mod)
+{
+	struct bt_mesh_lightness_srv *srv = mod->user_data;
+
+	srv->range.min = 0;
+	srv->range.max = UINT16_MAX;
+	srv->default_light = 0;
+	srv->last = UINT16_MAX;
+	atomic_clear_bit(&srv->flags, LIGHTNESS_SRV_FLAG_IS_ON);
+}
+
+static int bt_mesh_lightness_srv_init(struct bt_mesh_model *mod)
+{
+	struct bt_mesh_lightness_srv *srv = mod->user_data;
+
+	srv->lightness_model = mod;
+	bt_mesh_lightness_srv_reset(mod);
+	net_buf_simple_init(mod->pub->msg, 0);
+
+	if (IS_ENABLED(CONFIG_BT_MESH_MODEL_EXTENSIONS)) {
+		/* Model extensions:
+		 * To simplify the model extension tree, we're flipping the
+		 * relationship between the lightness server and the lightness
+		 * setup server. In the specification, the lightness setup
+		 * server extends the lightness server, which is the opposite of
+		 * what we're doing here. This makes no difference for the mesh
+		 * stack, but it makes it a lot easier to extend this model, as
+		 * we won't have to support multiple extenders.
+		 */
+		bt_mesh_model_extend(mod, srv->ponoff.ponoff_model);
+		bt_mesh_model_extend(
+			mod,
+			bt_mesh_model_find(
+				bt_mesh_model_elem(mod),
+				BT_MESH_MODEL_ID_LIGHT_LIGHTNESS_SETUP_SRV));
+	}
+
+	return 0;
+}
+
+#ifdef CONFIG_BT_SETTINGS
+static int bt_mesh_lightness_srv_settings_set(struct bt_mesh_model *mod,
+					      size_t len_rd,
+					      settings_read_cb read_cb,
+					      void *cb_arg)
+{
+	struct bt_mesh_lightness_srv *srv = mod->user_data;
+	struct bt_mesh_lightness_srv_settings_data data;
+
+	if (read_cb(cb_arg, &data, sizeof(data)) != sizeof(data)) {
```

Comment:
the callback returns
positive: Number of bytes read, 0: key-value pair is deleted. On error returns -ERRNO code.
https://docs.zephyrproject.org/latest/reference/runtime_conf/settings/settings.html#_CPPv416settings_read_cb